src/ir.h

```diff
@@ -144,70 +145,156 @@ struct Catch {
   ~Catch();
   Location loc;
   Var var;
-  struct Expr* first;
+  Expr* first;
   bool IsCatchAll() const {
     return var.type == VarType::Index && var.index == kInvalidIndex;
   }
 };
 
 typedef std::vector<Catch*> CatchVector;
 
-struct Expr {
+class Expr {
+ public:
   WABT_DISALLOW_COPY_AND_ASSIGN(Expr);
-  Expr();
-  explicit Expr(ExprType);
-  ~Expr();
-
-  static Expr* CreateBinary(Opcode);
-  static Expr* CreateBlock(Block*);
-  static Expr* CreateBr(Var);
-  static Expr* CreateBrIf(Var);
-  static Expr* CreateBrTable(VarVector* targets, Var default_target);
-  static Expr* CreateCall(Var);
-  static Expr* CreateCallIndirect(Var);
-  static Expr* CreateCompare(Opcode);
-  static Expr* CreateConst(const Const&);
-  static Expr* CreateConvert(Opcode);
-  static Expr* CreateCurrentMemory();
-  static Expr* CreateDrop();
-  static Expr* CreateGetGlobal(Var);
-  static Expr* CreateGetLocal(Var);
-  static Expr* CreateGrowMemory();
-  static Expr* CreateIf(Block* true_, Expr* false_ = nullptr);
-  static Expr* CreateLoad(Opcode, Address align, uint32_t offset);
-  static Expr* CreateLoop(Block*);
-  static Expr* CreateNop();
-  static Expr* CreateRethrow(Var);
-  static Expr* CreateReturn();
-  static Expr* CreateSelect();
-  static Expr* CreateSetGlobal(Var);
-  static Expr* CreateSetLocal(Var);
-  static Expr* CreateStore(Opcode, Address align, uint32_t offset);
-  static Expr* CreateTeeLocal(Var);
-  static Expr* CreateThrow(Var);
-  static Expr* CreateTry();
-  static Expr* CreateUnary(Opcode);
-  static Expr* CreateUnreachable();
+  Expr() = delete;
+  virtual ~Expr() {}
+
+  template <typename T>
+  bool Is() const {
+    WABT_STATIC_ASSERT((std::is_base_of<Expr, T>::value));
+    return type == T::kStaticType;
+  };
+
+  template <typename T>
+  const T* As() const {
+    assert(Is<T>());
+    return static_cast<const T*>(this);
+  }
+
+  template <typename T>
+  T* As() {
+    assert(Is<T>());
+    return static_cast<T*>(this);
+  }
```

Comment:
Yeah, that sounds like a good idea. We don't really have many type hierarchies yet, but my current plan is to use a bunch of them in `ir.h` so we don't have to deal with the union nastiness. So this will likely crop up more.

BTW, here is the doc describing the implementation: http://llvm.org/docs/HowToSetUpLLVMStyleRTTI.html